source/common/config/utility.cc

```diff
@@ -78,16 +78,34 @@ void Utility::checkFilesystemSubscriptionBackingPath(const std::string& path) {
   }
 }
 
-void Utility::checkApiConfigSourceSubscriptionBackingCluster(
-    const Upstream::ClusterManager::ClusterInfoMap& clusters,
+void Utility::checkApiConfigSourceNames(
     const envoy::api::v2::core::ApiConfigSource& api_config_source) {
-  if (api_config_source.cluster_names().size() != 1) {
-    // TODO(htuch): Add support for multiple clusters, #1170.
-    throw EnvoyException(
-        "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
+  const bool is_grpc =
+      (api_config_source.api_type() == envoy::api::v2::core::ApiConfigSource::GRPC);
+
+  if (is_grpc) {
+    if (api_config_source.cluster_names().size() != 0) {
+      ENVOY_LOG_MISC(
+          warn, "envoy::api::v2::core::ConfigSource::GRPC must not have a cluster name specified");
+    }
+    if (api_config_source.grpc_services().size() != 1) {
```

Comment:
Shouldn't this be an `else if` (how come the tests didn't catch this?).

Reply:
I think this shouldn't be an else if. The first block checks for non-zero cluster_names, which would be bad because setting a cluster name for a gRPC is deprecated. This second block independently checks for non-singleton grpc_services, which are unsupported.

Reply:
You'll need to change the check to `> 1` then, to support when `grpc_services.size() == 0` and they are using deprecated `cluster_names`.

Reply:
Oh, nice catch. We're sort of overloading `!=1` here to catch both empty grpc_services (deprecated) and `>1` grpc_services (not supported).